fecgov/fec-cms
# Improved Radform validation

## Summary (required)
- Improved client-side validation for [Radform](https://www.fec.gov/help-candidates-and-committees/question-rad/)
- Maintain server side validation should  there be an unforeseen server error or JS is not be available /fails on client-side
- Add JS test coverage
- A few particulars on issue checklist of user interaction/validation behavior has evolved/changed since the writing of the ticket. Please see add'l comments inn issue conversatino as well.
-  Resolves #4354
##  User interaction/ validation behavior:
- **Upon typing into a field**: User gets an error (red border and red message under field) once they focus away from a field they have typed in and is invalid 
    - Once they go back and type into an invalid field, the red error states go away and return if they focus away and the field is still invalid
    - The `Committee name or ID field` will error if a user does not choose from the Typeahead dropdown options
    - If user types in the `Committee name or ID field` after choosing from Typeahead options, the hidden ID field gets cleared and the Name or Id field remainns invalid until they choose from Typeahead again.
    - The recaptcha has its own validation built in as a user interacts with it before submitting
- **On Submit with invalid fields**: The page scrolls to an error message box listing the recaptcha error (if not checked) and a list of invalid fields  and all invalid fields are have red border and red message underneath
- **On Submit with no invalid  fields, but unchecked recaptca**: The page submits to the server and returns to the form with an error message indicating the recaptcha was not checked

- **- **On Submit with unforeseen server error or failing/disabled JS:** The user will get the server side validation, showing invalid fields similar to the the client side patterns.
       
See expected error messaging for message box at top of page in this issue-comment:
https://github.com/fecgov/fec-cms/issues/4354#issuecomment-951874330



### Required reviewers

One UX, one frontend, one content

## Impacted areas of the application

	modified:   fec/static/js/pages/contact-form.js
	modified:   fec/tests/js/radform-validate.js
	modified:   fec/tests/js/contact-form.js
	modified:   home/templates/home/contact-form.html
	modified:    static/scss/components/_contact-form.scss
	modified:   home/views.py

## Screenshots

![invalid_field](https://user-images.githubusercontent.com/5572856/138324873-1a61c09b-edca-46dc-85ad-4c040d96d8fd.gif)


## How to test
Test this on Feature :https://fec-feature-cms.app.cloud.gov/help-candidates-and-committees/question-rad/
--OR--
- checkout branch
- Export credentials for Service now and Recaptcha in your shell or bash_profile.
- npm run build
- View page at http://localhost:8000/help-candidates-and-committees/question-rad/
    - must use localhost, not 127.0.... for Recaptcha to work
    - The form will submit to Rad if you have all the creds exported. Brad said it's fine to submit tests, but make sure you put something in the text field to let them know its a test if you plan on submitting a fully valid form.
- If you want to see the server side validation, you could:
    - comment out these two lines from `contact-form.js` and npm run build-js to and restart server
    `new RadFormValidate('#id_contact_form');`
    `RadFormValidate: RadFormValidate`
- Also run `npm run test-single` to see that tests pass
